### PR TITLE
Update Rust in CI to 1.72.0, clarify Wasmtime's MSRV

### PR DESCRIPTION
**Pull request by alexcrichton**
This PR updates Rust in CI to 1.72.0. Additionally this implements the outcome of the latest Wasmtime meeting's discussion about an MSRV policy which is to support the latest 3 releases of Rust. CI now primarily tests against the latest release of Rust but a single Linux x86_64 builder runs the full test suite on the minimum Rust version supported.

### CODE REVIEW

**Comment by cfallin** [comment]
I'll offer a point in favor of option 2, with uncertain weight but maybe nonzero: not everyone uses `rustup` and can easily update their Rust compiler right away. Probably most folks do, and it's certainly the recommended way to use Rust, but the language is also packaged in distros, installed system-wide on shared machines, etc. Also platforms that are Rust tier 3 (hence have no rustup binaries) but have OS-packaged rustc and happen to work (at one point this included various BSDs, especially non-x86). Finally, some projects (e.g. Firefox back when it had an option to use Cranelift) pin on a certain Rust version, and so we're excluding ourselves from use in large software systems that may have more involved Rust upgrade procedures. (Any large company with a monorepo and carefully-packaged toolchains, for example.)

Many of those probably carry lower weight than e.g. using a sorely-needed feature in a new release; but if the habit is instead to update whenever we have time, with no other particular benefits, maybe we could consider the above?

**Comment by elliottt** [comment]
I'm generally in favor of staying up-to-date. What about adopting policy 1, and re-examining it if we get a report that it's breaking someone else's workflow?

**Comment by cfallin** [comment]
Ah, sorry, I didn't mean to reignite any old discussions or scars. AFAIK this is the first time I've been involved in a "MSRV discussion"; hopefully I don't have as bumpy a path with them!

I agree it'd be great to set down an explicit policy for Wasmtime, to avoid any ambiguity. Maybe we should discuss and decide on something in the next Wasmtime meeting? I'll put it on the agenda.

**Comment by cfallin** [comment]
(My comment raced with @elliottt's; I'm fine with seeing this go through as-is, then we can discuss more at the meeting...)

**Review by elliottt — APPROVED** [review]

**Comment by alexcrichton** [comment]
Ok sounds good! I'm happy to sit on this until the next meeting. If someone wants to use something from 1.72 I'd be in favor of merging and enabling that, but in the meantime I think it's ok to update locally and otherwise keep CI at 1.71

**Comment by cfallin** [comment]
We talked about this in the Wasmtime biweekly just now, and reached a good amount of consensus; to summarize: we agreed that it makes sense to support a sliding window of latest stable Rust versions (stable down to stable - N; N is a parameter we still need to choose). This allows flexibility to embedders of Wasmtime by not forcing an upgrade to latest stable immediately after release -- others can design release processes that work within the N-version window. It also gives some flexibility for embedders if they ned to hold a version upgrade temporarily for whatever reason (bug or miscompile elsewhere, etc).

The main points to be decided are:
* What is N? (Some common choices seem to be 1, 2, 4; ideally not too much more than that?)
* How do we test this in CI?

On the latter point, discussion revolved around (at least) two main constraints: CI resources are limited, so that we can't simply duplicate the test runs for all Rust versions in the window; and we also would prefer "CI determinism", i.e., a PR that previously passed CI should not stop passing if a new Rust version happens to be released between runs. (This is important for the contributor experience but also for e.g. ability to respond quickly without roadblocks to emergency patches and that sort of thing.)

Given the determinism constraint, we'd likely define the window by pinning two versions of rust -- "newest supported" and "oldest supported" -- defining our window manually and bumping it manually. We should do these bumps regularly when new releases of Rust occur, and *not* only when we actually need a feature: otherwise, compatibility pressure builds up, dependencies acrete, and we have more problems when we do bump.

Given the CI resources constraint, there seem to be two main options: run all tests at newest supported and add one additional job on Linux/x86-64 that runs at oldest supported, or vice-versa. The former catches platform-specific issues due to new Rust features sooner, but does not cover the "tail" of the sliding window, so N-1/N-2 Rust could break in platform-specific ways. The latter ensures the tail of the window stays working everywhere, but may catch platform-specific issues later.

Does this capture everything @alexcrichton, @fitzgen, @tschneidereit, others?

---

My own thoughts:
* I think `N = 2` strikes a reasonable balance; it gives a 12-week window for upgrades if someone is tracking latest Wasmtime.
* It seems to me that emprically, it may be better to pin most platform CI jobs to newest supported Rust, and cover the tail of the window with one Linux/x86-64 job: we've had more issues from new Rust versions (miscompiles mainly) than issues from changes to Wasmtime interacting poorly with older Rust. (Or said another way, the Rust compiler makes large changes more frequently than Wasmtime, so let's test new Rust everywhere, and new Wasmtime features/PRs only on one older Rust.) But I could see us going either way here.

**Comment by alexcrichton** [comment]
Thanks for writing that up @cfallin and that all looks accurate to me! I'd also agree with your conclusions of N=2 and "test latest, one job msrv" personally.

File: docs/contributing-coding-guidelines.md [hunk 1]
```diff
@@ -21,13 +21,13 @@ your editor.
 
 ### Minimum Supported `rustc` Version
 
-Wasmtime supports the current stable version of Rust.
+Wasmtime supports the latest three stable releases of Rust.
 
 Cranelift supports stable Rust, and follows the [Rust Update Policy for
```
Comment (cfallin) [inline docs/contributing-coding-guidelines.md:26]: While we're here, could we update this for Cranelift as well? We're no longer included in the mozilla-central tree and aren't subject to their Rust version policies anymore; it's probably uncontroversial to say now that Cranelift will follow the same policy as Wasmtime?

**Review by cfallin — APPROVED** [review]
LGTM for latest updates; thanks!

**Comment by fitzgen** [comment]
> Thanks for writing that up @cfallin and that all looks accurate to me! I'd also agree with your conclusions of N=2 and "test latest, one job msrv" personally.

+1 on both these things from me as well.

**Comment by tschneidereit** [comment]
> > Thanks for writing that up @cfallin and that all looks accurate to me! I'd also agree with your conclusions of N=2 and "test latest, one job msrv" personally.
> 
> +1 on both these things from me as well.

And from me, thank you!

**Comment by alexcrichton** [comment]
Ok I've updated the OP and title of this PR as well, thanks again everyone for all the input 👍

**Pull request merged**